src/Autofac/Core/ServiceRegistration.cs

```diff
@@ -0,0 +1,98 @@
+﻿// This software is part of the Autofac IoC container
+// Copyright © 2020 Autofac Contributors
+// https://autofac.org
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Collections.Generic;
+using Autofac.Builder;
+using Autofac.Core.Resolving.Pipeline;
+
+namespace Autofac.Core
+{
+    /// <summary>
+    /// Defines a combination of a service pipeline and a registration. Used to instantiate a <see cref="ResolveRequest"/>.
+    /// </summary>
+    public struct ServiceRegistration : IEquatable<ServiceRegistration>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServiceRegistration"/> struct.
+        /// </summary>
+        /// <param name="servicePipeline">A service pipeline.</param>
+        /// <param name="registration">The registration.</param>
+        public ServiceRegistration(IResolvePipeline servicePipeline, IComponentRegistration registration)
+        {
+            Pipeline = servicePipeline;
```

Comment:
If these are null it'll break a bunch of stuff. Admittedly, chances are super slim, but... yeah.